enterprise/server/raft/cache/cache.go

```diff
@@ -541,73 +466,78 @@ func (rc *RaftCache) findMissingResourceNames(ctx context.Context, resourceNames
 	}
 
-	var allMissingDigests []*repb.Digest
+	// Combine the partial responses from each shard.
+	allFound := make(map[*sgpb.FileRecord]*sgpb.FileMetadata, len(req.GetFileRecords()))
 	for _, rsp := range rsps {
-		res, ok := rsp.(findResult)
+		res, ok := rsp.(getMetadataResult)
 		if !ok {
-			return nil, status.InternalError("response not of type findResult")
+			return nil, status.InternalError("response not of type getMetadataResult")
+		}
+
+		for k, v := range res.found {
+			allFound[k] = v
 		}
 
 		for _, p := range res.atimeUpdates {
 			rc.sendAccessTimeUpdate(p.key, p.lastAccessUsec)
 		}
-
-		allMissingDigests = append(allMissingDigests, res.missingDigests...)
 	}
 
-	return allMissingDigests, nil
-}
-
-func (rc *RaftCache) FindMissing(ctx context.Context, resources []*rspb.ResourceName) ([]*repb.Digest, error) {
-	return rc.findMissingResourceNames(ctx, resources)
+	// Assemble the response proto.
+	rsp := &mdpb.GetResponse{
+		FileMetadatas: make([]*sgpb.FileMetadata, len(req.GetFileRecords())),
+	}
+	for i, fileRecord := range req.GetFileRecords() {
+		rsp.FileMetadatas[i] = allFound[fileRecord]
+	}
+	return rsp, nil
 }
 
-func (rc *RaftCache) Get(ctx context.Context, rn *rspb.ResourceName) ([]byte, error) {
-	r, err := rc.Reader(ctx, rn, 0, 0)
-	if err != nil {
-		return nil, err
-	}
-	defer r.Close()
-	return io.ReadAll(r)
+type findMetadataResult struct {
+	found        map[*sgpb.FileRecord]bool
+	atimeUpdates []keyAndLastAccessUsec
 }
 
-func (rc *RaftCache) GetMulti(ctx context.Context, resources []*rspb.ResourceName) (map[*repb.Digest][]byte, error) {
-	keys, err := rc.resourceNamesToKeyMetas(ctx, resources)
+func (rc *RaftCache) Find(ctx context.Context, req *mdpb.FindRequest) (*mdpb.FindResponse, error) {
+	keys, err := rc.fileRecordsToKeyMetas(req.GetFileRecords())
 	if err != nil {
 		return nil, err
 	}
 
+	// Shard the query by key and query shards in parallel.
 	rsps, err := rc.sender().RunMultiKey(ctx, keys, func(c rfspb.ApiClient, h *rfpb.Header, keys []*sender.KeyMeta) (interface{}, error) {
 		batch := rbuilder.NewBatchBuilder()
 		for _, k := range keys {
-			batch.Add(&rfpb.GetRequest{
+			batch.Add(&rfpb.FindRequest{
 				Key: k.Key,
 			})
 		}
 		batchProto, err := batch.ToProto()
 		if err != nil {
 			return nil, err
 		}
-		// TODO(tylerw): add a SyncReadMulti?
 		rsp, err := c.SyncRead(ctx, &rfpb.SyncReadRequest{
 			Header: h,
 			Batch:  batchProto,
 		})
 		if err != nil {
 			return nil, err
 		}
-		res := getMultiResult{
-			found: make(map[*repb.Digest][]byte),
+		res := findMetadataResult{
+			found: make(map[*sgpb.FileRecord]bool),
 		}
 		batchRsp := rbuilder.NewBatchResponseFromProto(rsp.GetBatch())
 		for i, k := range keys {
-			r, err := batchRsp.GetResponse(i)
-			if err == nil {
-				fr := k.Meta.(*sgpb.FileRecord)
-				res.found[fr.GetDigest()] = r.GetFileMetadata().GetStorageMetadata().GetInlineMetadata().GetData()
+			findRsp, err := batchRsp.FindResponse(i)
+			if err != nil {
```

Comment:
I think findResponse returns NotFoundError so you might not want to return if error .